src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts

```diff
@@ -305,4 +297,104 @@ export class SettingMatches {
 	}
 }
 
+class RemoteSearchProviderKeyCache {
+	private settingKeys: string[] = [];
+	private settingsRecord: Record<string, ISetting> = {};
+	private currentPreferencesModel: ISettingsEditorModel | undefined;
+
+	constructor(private readonly semanticSimilarityService: ISemanticSimilarityService) {
+	}
+
+	updateModel(preferencesModel: ISettingsEditorModel) {
+		if (preferencesModel === this.currentPreferencesModel) {
+			return;
+		}
+
+		this.currentPreferencesModel = preferencesModel;
+		this.refresh();
+	}
+
+	refresh() {
+		this.settingKeys = [];
+		this.settingsRecord = {};
+
+		if (!this.semanticSimilarityService.isEnabled() || !this.currentPreferencesModel) {
+			return;
+		}
+
+		for (const group of this.currentPreferencesModel.settingsGroups) {
+			for (const section of group.sections) {
+				for (const setting of section.settings) {
+					this.settingKeys.push(setting.key);
+					this.settingsRecord[setting.key] = setting;
+				}
+			}
+		}
+	}
+
+	getSettingKeys(): string[] {
+		return this.settingKeys;
+	}
+
+	getSettingsRecord(): Record<string, ISetting> {
+		return this.settingsRecord;
+	}
+}
+
+export class RemoteSearchProvider implements ISearchProvider {
+	private static readonly SEMANTIC_SIMILARITY_THRESHOLD = 0.75;
+	private static readonly SEMANTIC_SIMILARITY_MAX_PICKS = 10;
+
+	private readonly _keysProvider: RemoteSearchProviderKeyCache;
+	private _filter: string = '';
+
+	constructor(
+		@ISemanticSimilarityService private readonly semanticSimilarityService: ISemanticSimilarityService,
+	) {
+		this._keysProvider = new RemoteSearchProviderKeyCache(semanticSimilarityService);
+	}
+
+	setFilter(filter: string) {
+		this._filter = cleanFilter(filter);
+	}
+
+	searchModel(preferencesModel: ISettingsEditorModel, token?: CancellationToken | undefined): Promise<ISearchResult | null> {
+		if (!this.semanticSimilarityService.isEnabled() || !this._filter) {
+			return Promise.resolve(null);
+		}
+
+		this._keysProvider.updateModel(preferencesModel);
+		const settingKeys = this._keysProvider.getSettingKeys();
+		const settingsRecord = this._keysProvider.getSettingsRecord();
+
+		return this.semanticSimilarityService.getSimilarityScore(this._filter, settingKeys, token ?? CancellationToken.None).then((scores) => {
+			const filterMatches: ISettingMatch[] = [];
+			const foundKeys = new Set<string>();
```

Comment:
Something for me to confirm is whether the current list of keys contains duplicates.